src/server.rs

```diff
@@ -0,0 +1,189 @@
+    KeepAlive,          // CH
+    AskListLengths,     // askchaa
+    AskListCharacters,  // askchar
+    CharacterList(u32),      // AN#<page:u32>#%
+    EvidenceList(u32),       // AE#<page:u32>#%
+    MusicList(u32),          // AM#<page:u32>#%
+    AO2CharacterList,   // AC#%
+    AO2MusicList,       // AM#%
+    AO2Ready,           // RD#%
+    SelectCharacter(u32, u32, String),    // CC<client_id:u32>#<char_id:u32#<hdid:String>#%
+    ICMessage,          // MS
+    OOCMessage(String, String),         // CT#<name:String>#<message:String>#%
+    PlaySong(u32, u32),           // MC#<song_name:u32>#<???:u32>#%
+    WTCEButtons(String),        // RT#<type:String>#%
+    SetCasePreferences(String, CasePreferences), // SETCASE#<cases:String>#<will_cm:boolean>#<will_def:boolean>#<will_pro:boolean>#<will_judge:boolean>#<will_jury:boolean>#<will_steno:boolean>#%
+    CaseAnnounce(String, CasePreferences),       // CASEA
+    Penalties(u32, u32),          // HP#<type:u32>#<new_value:u32>#%
+    AddEvidence(EvidenceArgs),        // PE#<name:String>#<description:String>#<image:String>#%
+    DeleteEvidence(u32),     // DE#<id:u32>#%
+    EditEvidence(u32, EvidenceArgs),       // EE#<id:u32>#<name:String>#<description:String>#<image:String>#%
+    CallModButton(Option<String>),      // ZZ?#<reason:String>?#%
+}
+
+impl Command for ClientCommand {
+    fn from_protocol(name: String, mut args: Vec<String>) -> Result<Self, anyhow::Error> {
+        let args_len = args.len();
+        match name.as_str() {
+            "HI" => {
+                if args_len != 1 {
+                    anyhow::bail!("Amount of arguments for command HANDSHAKE does not match!")
+                }
+
+                Ok(Self::Handshake(args.remove(0)))
+            },
+            _ => Ok(Self::ICMessage)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EvidenceArgs {
+    pub name: String,
+    pub description: String,
+    pub image: String,
+}
+
+#[derive(Debug)]
+pub struct CasePreferences {
+    pub cm: bool,
+    pub def: bool,
+    pub pro: bool,
+    pub judge: bool,
+    pub jury: bool,
+    pub steno: bool
+}
+
+#[derive(Debug)]
+pub struct AOMessage {
+    pub command: ClientCommand,
+}
+
+pub struct AOMessageCodec;
+
+impl Decoder for AOMessageCodec {
+    type Item = AOMessage;
+    type Error = anyhow::Error;
+
+    fn decode(
+        &mut self,
+        mut src: &mut BytesMut,
+    ) -> Result<Option<Self::Item>, Self::Error> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+        let mut reader = Cursor::new(&src);
+        let mut cmd_buf = Vec::with_capacity(8);
+        let cmd_len = BufRead::read_until(&mut reader, b'#', &mut cmd_buf)?;
+
+        if cmd_len == 0 {
+            log::error!("Invalid protocol?!");
+            anyhow::bail!("Invalid protocol!")
+        }
+
+        cmd_buf.truncate(cmd_len.saturating_sub(1));
+        let command_name = String::from_utf8_lossy(&cmd_buf).replace("�", "");
```

Comment:
`replace` actually converts to `String` always, so it's better to call it only on invalid utf8:
```suggestion
        let command_name = match String::from_utf8_lossy(&cmd_buf) {
            b @ Cow::Borrowed(_) => b,
            Cow::Owned(string) => string.replace("�", ""),
        }
```

Reply:
Also, `retain` could be used: `Cow::Owned(mut string) => { string.retain(|c| c == '�'); string },`